agent/provider/src/cli/whitelist.rs

```diff
@@ -0,0 +1,200 @@
+use std::collections::HashMap;
+
+use structopt::StructOpt;
+
+use ya_manifest_utils::{
+    matching::domain::{pattern_to_id, DomainPattern, DomainPatterns},
+    ArgMatch,
+};
+use ya_utils_cli::{CommandOutput, ResponseTable};
+
+use crate::cli::println_conditional;
+use crate::startup_config::ProviderConfig;
+
+#[derive(StructOpt, Clone, Debug)]
+#[structopt(rename_all = "kebab-case")]
+pub enum WhitelistConfig {
+    /// List domain whitelist patterns
+    List,
+    /// Add new domain whitelist patterns
+    Add(Add),
+    /// Remove domain whitelist patterns
+    Remove(Remove),
+}
+
+#[derive(StructOpt, Clone, Debug)]
+#[structopt(rename_all = "kebab-case")]
+pub struct Add {
+    /// Domain whitelist patterns
+    #[structopt(long, short)]
+    patterns: Vec<String>,
+
+    /// Domain whitelist patterns type
+    #[structopt(long = "type", short = "t", default_value = "strict")]
+    pattern_type: ArgMatch,
+}
+
+#[derive(StructOpt, Clone, Debug)]
+#[structopt(rename_all = "kebab-case")]
+pub struct Remove {
+    /// Domain whitelist pattern ids.
+    ids: Vec<String>,
+}
+
+impl WhitelistConfig {
+    pub fn run(self, config: ProviderConfig) -> anyhow::Result<()> {
+        match self {
+            WhitelistConfig::List => list(config),
+            WhitelistConfig::Add(cmd) => add(config, cmd),
+            WhitelistConfig::Remove(cmd) => remove(config, cmd),
+        }
+    }
+}
+
+fn list(config: ProviderConfig) -> anyhow::Result<()> {
+    let domain_patterns = DomainPatterns::load_or_create(&config.domain_whitelist_file)?;
+    let table = WhitelistTable::from(domain_patterns);
+    table.print(&config)
+}
+
+fn add(config: ProviderConfig, add: Add) -> anyhow::Result<()> {
+    let domain_patterns = DomainPatterns::load_or_create(&config.domain_whitelist_file)?;
+    let mut domain_patterns = DomainPatternIds::from(domain_patterns);
+    let (added, skipped) = domain_patterns.add(add);
+    let domain_patterns: DomainPatterns = domain_patterns.into();
+    domain_patterns.save(&config.domain_whitelist_file)?;
+    if !added.is_empty() {
+        println_conditional(&config, "Added patterns:");
+        WhitelistTable::from(DomainPatterns { patterns: added }).print(&config)?
+    } else if config.json {
+        println_conditional(&config, "No new patterns to add.");
+        // no new pattern to add, so empty list for json output
+        WhitelistTable::from(DomainPatterns {
+            patterns: Vec::new(),
+        })
+        .print(&config)?
+    }
+    if !skipped.is_empty() && !config.json {
+        println_conditional(&config, "Dropped duplicated patterns:");
+        if !config.json {
+            WhitelistTable::from(DomainPatterns { patterns: skipped }).print(&config)?
+        }
```

Comment:
Redundant logic:
* `println_conditional` will always print.
* The inner `if !config.json` will always execute
Both are because you already assert `config.json` in the outer if expression.

I suppose you could be using `println_conditional` everywhere for consistency, that would be fine. All in all I would prefer normal `println!` when possible to clear up the logic, but those issues won't block my approval.

Reply:
Will change this println. A consistency was the reason for using it.
Second `if` was a shameful mistake (again on `skipped`..).
Done.